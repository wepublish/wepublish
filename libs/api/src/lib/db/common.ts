import {PageInfo} from '@wepublish/utils/api'

export interface ConnectionResult<T> {
  nodes: T[]
  pageInfo: PageInfo
  totalCount: number
}

export const DefaultSessionTTL = 1000 * 60 * 60 * 24 * 7 // 1w
export const DefaultBcryptHashCostFactor = 11

export enum DateFilterComparison {
  GreaterThan = 'gt',
  GreaterThanOrEqual = 'gte',
  Equal = 'eq',
  LowerThan = 'lt',
  LowerThanOrEqual = 'lte'
}
export interface DateFilter {
  date: Date | null
  comparison: DateFilterComparison
}
