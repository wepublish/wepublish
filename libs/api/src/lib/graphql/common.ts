import { DateFilterComparison, SortOrder } from '@wepublish/utils/api';
import {
  GraphQLBoolean,
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from 'graphql';
import { GraphQLDateTime } from 'graphql-scalars';

export const GraphQLSortOrder = new GraphQLEnumType({
  name: 'SortOrder',
  values: {
    [SortOrder.Ascending]: { value: SortOrder.Ascending },
    [SortOrder.Descending]: { value: SortOrder.Descending },
  },
});

export const GraphQLPageInfo = new GraphQLObjectType({
  name: 'PageInfo',
  fields: {
    startCursor: { type: GraphQLString },
    endCursor: { type: GraphQLString },
    hasNextPage: { type: new GraphQLNonNull(GraphQLBoolean) },
    hasPreviousPage: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLUnidirectionalPageInfo = new GraphQLObjectType({
  name: 'UnidirectionalPageInfo',
  fields: {
    endCursor: { type: GraphQLString },
    hasNextPage: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataProperty = new GraphQLObjectType({
  name: 'Properties',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
    public: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataPropertyPublic = new GraphQLObjectType({
  name: 'PublicProperties',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
  },
});

export const GraphQLMetadataPropertyInput = new GraphQLInputObjectType({
  name: 'PropertiesInput',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
    public: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataPropertyPublicInput = new GraphQLInputObjectType({
  name: 'PublicPropertiesInput',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
  },
});

export const GraphQLDateFilterComparison = new GraphQLEnumType({
  name: 'DateFilterComparison',
  values: {
    [DateFilterComparison.GreaterThan]: {
      value: DateFilterComparison.GreaterThan,
    },
    [DateFilterComparison.GreaterThanOrEqual]: {
      value: DateFilterComparison.GreaterThanOrEqual,
    },
    [DateFilterComparison.Equal]: { value: DateFilterComparison.Equal },
    [DateFilterComparison.LowerThan]: { value: DateFilterComparison.LowerThan },
    [DateFilterComparison.LowerThanOrEqual]: {
      value: DateFilterComparison.LowerThanOrEqual,
    },
  },
});

export const GraphQLDateFilter = new GraphQLInputObjectType({
  name: 'DateFilter',
  fields: {
    date: { type: GraphQLDateTime, defaultValue: null },
    comparison: { type: new GraphQLNonNull(GraphQLDateFilterComparison) },
  },
});
