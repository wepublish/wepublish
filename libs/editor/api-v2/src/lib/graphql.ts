// THIS FILE IS AUTOGENERATED
import {Node} from 'slate'
import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
const defaultOptions = {} as const;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  versionInformation: VersionInformation;
};

export type VersionInformation = {
  __typename?: 'VersionInformation';
  version: Scalars['String'];
};

export type VersionInformationQueryVariables = Exact<{ [key: string]: never; }>;


export type VersionInformationQuery = { __typename?: 'Query', versionInformation: { __typename?: 'VersionInformation', version: string } };


export const VersionInformationDocument = gql`
    query VersionInformation {
  versionInformation {
    version
  }
}
    `;

/**
 * __useVersionInformationQuery__
 *
 * To run a query within a React component, call `useVersionInformationQuery` and pass it any options that fit your needs.
 * When your component renders, `useVersionInformationQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useVersionInformationQuery({
 *   variables: {
 *   },
 * });
 */
export function useVersionInformationQuery(baseOptions?: Apollo.QueryHookOptions<VersionInformationQuery, VersionInformationQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<VersionInformationQuery, VersionInformationQueryVariables>(VersionInformationDocument, options);
      }
export function useVersionInformationLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<VersionInformationQuery, VersionInformationQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<VersionInformationQuery, VersionInformationQueryVariables>(VersionInformationDocument, options);
        }
export type VersionInformationQueryHookResult = ReturnType<typeof useVersionInformationQuery>;
export type VersionInformationLazyQueryHookResult = ReturnType<typeof useVersionInformationLazyQuery>;
export type VersionInformationQueryResult = Apollo.QueryResult<VersionInformationQuery, VersionInformationQueryVariables>;