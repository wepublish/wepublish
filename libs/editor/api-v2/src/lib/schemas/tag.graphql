fragment FullTag on Tag {
  id
  tag
  description
  type
  main
  url
  color
}
