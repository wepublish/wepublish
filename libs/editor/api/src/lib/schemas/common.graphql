

fragment MetadataProperty on Properties {
    key
    value
    public
}
