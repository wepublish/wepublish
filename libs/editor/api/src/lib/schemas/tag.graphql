query TagList(
  $filter: TagFilter
  $cursor: ID
  $take: Int
  $skip: Int
  $order: SortOrder
  $sort: TagSort
) {
  tags(filter: $filter, cursor: $cursor, take: $take, skip: $skip, order: $order, sort: $sort) {
    nodes {
      id
      tag
      main
    }

    pageInfo {
      startCursor
      endCursor
      hasNextPage
      hasPreviousPage
    }

    totalCount
  }
}

mutation CreateTag($tag: String, $type: TagType!) {
  createTag(tag: $tag, type: $type) {
    id
    tag
    main
  }
}

mutation UpdateTag($id: ID!, $tag: String, $main: Boolean) {
  updateTag(id: $id, tag: $tag, main: $main) {
    id
    tag
    main
  }
}

mutation DeleteTag($id: ID!) {
  deleteTag(id: $id) {
    id
    tag
    main
  }
}
