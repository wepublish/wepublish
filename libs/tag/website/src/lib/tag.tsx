import styled from '@emotion/styled';
import {
  BuilderTagProps,
  TagSEO,
  useWebsiteBuilder,
} from '@wepublish/website/builder';
import { ContentWrapper } from '@wepublish/content/website';
import { ArticleListWrapper } from '@wepublish/article/website';
import { useMemo } from 'react';
import { capitalize } from '@mui/material';

export const TagWrapper = styled(ContentWrapper)`
  ${({ theme }) => theme.breakpoints.up('md')} {
    & > :is(${ArticleListWrapper}) {
      grid-column: 2/12;
    }
  }
`;

export const TagTitle = styled('h1')``;

export const TagTitleWrapper = styled('div')`
  display: grid;
  gap: ${({ theme }) => theme.spacing(3)};
  grid-auto-rows: min-content;
`;

export function Tag({
  className,
  tags,
  articles,
  variables,
  onVariablesChange,
}: BuilderTagProps) {
  const {
    ArticleList,
    elements: { Alert, Pagination, H2 },
    blocks: { RichText },
  } = useWebsiteBuilder();

  const tag = tags.data?.tags?.nodes.at(0);
  const take = variables?.take ?? 1;
  const page = variables?.skip ? variables.skip / take + 1 : 1;

  const pageCount = useMemo(() => {
    if (
      articles.data?.articles?.totalCount &&
      articles.data.articles.totalCount > take
    ) {
      return Math.ceil(articles.data.articles.totalCount / take);
    }

    return 1;
  }, [articles.data?.articles?.totalCount, take]);

  if (!tag) {
    return;
  }

  return (
    <TagWrapper className={className}>
      <TagSEO tag={tag} />

      <TagTitleWrapper>
        <H2 component={TagTitle}>{capitalize(tag.tag ?? '')}</H2>

        {!!tag.description?.length && <RichText richText={tag.description} />}
      </TagTitleWrapper>

      {!articles.loading && !articles.data?.articles?.nodes.length && (
        <Alert severity="info">Keine Artikel vorhanden</Alert>
      )}

      <ArticleList
        {...articles}
        variables={variables}
        onVariablesChange={onVariablesChange}
      />

      {pageCount > 1 && (
        <Pagination
          page={page ?? 1}
          count={pageCount}
          onChange={(_, value) =>
            onVariablesChange?.({
              skip: (value - 1) * take,
            })
          }
        />
      )}
    </TagWrapper>
  );
}
