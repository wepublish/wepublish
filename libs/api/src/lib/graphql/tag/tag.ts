import {Tag, TagType} from '@prisma/client'
import {
  GraphQLEnumType,
  GraphQLID,
  GraphQLInputObjectType,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString
} from 'graphql'
import {Context} from '../../context'
import {GraphQLPageInfo} from '../common'
import {TagSort} from './tag.query'

export const GraphQLTagType = new GraphQLEnumType({
  name: 'TagType',
  values: {
    Comment: {value: TagType.Comment},
    Event: {value: TagType.Event}
  }
})

export const GraphQLTag = new GraphQLObjectType<Tag, Context>({
  name: 'Tag',
  fields: {
    id: {type: GraphQLNonNull(GraphQLID)},
    tag: {type: GraphQLString},
    type: {type: GraphQLTagType}
  }
})

export const GraphQLTagConnection = new GraphQLObjectType({
  name: 'TagConnection',
  fields: {
    nodes: {type: GraphQLNonNull(GraphQLList(GraphQLNonNull(GraphQLTag)))},
    pageInfo: {type: GraphQLNonNull(GraphQLPageInfo)},
    totalCount: {type: GraphQLNonNull(GraphQLInt)}
  }
})

export const GraphQLTagFilter = new GraphQLInputObjectType({
  name: 'TagFilter',
  fields: {
    type: {type: GraphQLTagType},
    tag: {type: GraphQLString}
  }
})

export const GraphQLTagSort = new GraphQLEnumType({
  name: 'TagSort',
  values: {
    CREATED_AT: {value: TagSort.CreatedAt},
    MODIFIED_AT: {value: TagSort.ModifiedAt},
    TAG: {value: TagSort.Tag}
  }
})
