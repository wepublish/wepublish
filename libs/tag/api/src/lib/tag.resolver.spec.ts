import {INestApplication} from '@nestjs/common'
import {createMock, PartialMocked, TagList} from '@wepublish/testing'
import {Test, TestingModule} from '@nestjs/testing'
import {GraphQLModule} from '@nestjs/graphql'
import {ApolloDriver, ApolloDriverConfig} from '@nestjs/apollo'
import {URLAdapter} from '@wepublish/nest-modules'
import {TagResolver} from './tag.resolver'
import {TagService} from './tag.service'
import request from 'supertest'

describe('TagResolver', () => {
  let app: INestApplication
  let tagService: PartialMocked<TagService>
  let urlAdapter: PartialMocked<URLAdapter>

  beforeEach(async () => {
    tagService = createMock(TagService)
    urlAdapter = createMock(URLAdapter)

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        GraphQLModule.forRoot<ApolloDriverConfig>({
          driver: ApolloDriver,
          autoSchemaFile: true,
          path: '/',
          cache: 'bounded'
        })
      ],
      providers: [
        TagResolver,
        {
          provide: TagService,
          useValue: tagService
        },
        {
          provide: URLAdapter,
          useValue: urlAdapter
        }
      ]
    }).compile()

    app = module.createNestApplication()
    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  test('can query a list of tags', async () => {
    urlAdapter.getTagURL?.mockResolvedValue('url-id1')
    tagService.getTags?.mockResolvedValue({
      nodes: [
        {
          id: 'id1',
          tag: 'tag1'
        }
      ],
      pageInfo: {
        startCursor: 'startCursor',
        endCursor: 'endCursor',
        hasNextPage: false,
        hasPreviousPage: false
      },
      totalCount: 0
    } as any)

    await request(app.getHttpServer())
      .post('')
      .send({
        query: TagList,
        variables: {
          filter: {
            tag: 'tag1'
          }
        }
      })
      .expect(response => {
        expect(tagService.getTags?.mock.calls).toMatchSnapshot()
        expect(urlAdapter.getTagURL?.mock.calls).toMatchSnapshot()
        expect(response.body.errors).toBeUndefined()
        expect(response.body.data).toMatchSnapshot()
      })
  })
})
