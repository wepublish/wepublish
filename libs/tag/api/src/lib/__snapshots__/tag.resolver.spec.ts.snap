// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagResolver can query a list of tags 1`] = `
[
  [
    {
      "tag": "tag1",
    },
    "CreatedAt",
    "Descending",
    undefined,
    0,
    10,
  ],
]
`;

exports[`TagResolver can query a list of tags 2`] = `
[
  [
    {
      "createdAt": 2023-01-01T00:00:00.000Z,
      "description": [],
      "id": "id1",
      "main": false,
      "modifiedAt": 2023-01-01T00:00:00.000Z,
      "peerId": null,
      "tag": "tag1",
      "type": "Article",
    },
  ],
]
`;

exports[`TagResolver can query a list of tags 3`] = `
{
  "tags": {
    "nodes": [
      {
        "id": "id1",
        "tag": "tag1",
        "url": "url-id1",
      },
    ],
    "pageInfo": {
      "endCursor": "endCursor",
      "hasNextPage": false,
      "hasPreviousPage": false,
      "startCursor": "startCursor",
    },
    "totalCount": 0,
  },
}
`;
